### DISCUSSION TITLE
Seeking Advice on Component Styling with TailwindCSS

### DISCUSSION THREAD

**Discussion by henzelennard** [tailwindlabs/tailwindcss#12338] [Help] [answered]
Hello everyone,

I've recently transitioned more towards front-end development and have been exploring different ways to style the elements that are being rendered. However, I've yet to find a solution that fully satisfies my needs.

Initially, I used a React component library that offered pre-styled components. The styling could be chosen by passing a prop to the component. However, due to poor maintenance of this library, I switched to React-Bootstrap, which follows a similar approach.

Unfortunately, I encountered issues with implementing dark mode in React-Bootstrap. This led me to use a custom stylesheet, applying only the minimal stylesheet from React-Bootstrap.

My ultimate goal is to separate the appearance of a component from its context on the screen. For instance, let's say Team A provides five different types of buttons. Developers in Team B can choose from these buttons by applying a className like "button-1" to "button-5". They can then modify how this button is displayed in relation to other components by adding properties like marginLeft: 1rem, etc.

I'm curious to know if this is a reasonable expectation or if I need to adjust my mental model for thinking about designs. Is it possible to achieve this result with TailwindCSS, or would I be better off using a different tool?

I appreciate any feedback or advice you can provide. Thank you!

**Comment by alex-krasikau** [comment, marked as answer]
Hey, @henzelennard!

Yes, it's possible to achieve that with Tailwind CSS. Here's a common way to do it:

```js
// Create a button component and include a style prop, commonly called `variant`.
const Button = ({ className, variant, ...buttonProps }: { className?: string, variant: "primary" | "secondary" /*... button props*/ }) => {
  return <button className={/* set the correct class based on the variant property */ + " " + className} {...buttonProps} />
}

// Then, you can use this component and apply "layout" classes to control its position:
<Button className="mt-2" />
// or
<Button className="absolute top-4 left-4" />
```

To enhance the solution for cleanliness and flexibility, I recommend considering the following libraries:
- [CVA](https://cva.style/docs) - for easily managing multiple variant options, such as variant (`primary | secondary`), color (`regular | accent`), and size (`small | medium | large`).
- [clsx](https://github.com/lukeed/clsx) - to merge the classes - `className("common-button", { "primary-button": variant === "primary" }, className)`.
- [Tailwind Merge](https://github.com/dcastil/tailwind-merge) - to deduplicate Tailwind classes when you want to override any of the internal component styles, such as border, color, or background.

**Reply by henzelennard** [reply]
Hi @alex-krasikau , thanks for your helpful advice! I'm really interested in your suggestion to use CVA. It seems like a great way to make clear, reusable component variants together with Tailwind!

**Reply by ankursehdev** [reply]
You can also use [Tailwind Variants](https://www.tailwind-variants.org/), I find it intuitive and faster than CVA
